core/src/utils/prove_distributed/checkpoints.rs

```diff
@@ -0,0 +1,154 @@
+            for record in records.iter_mut() {
+                state.shard += 1;
+                state.execution_shard = record.public_values.execution_shard;
+                state.start_pc = record.public_values.start_pc;
+                state.next_pc = record.public_values.next_pc;
+                record.public_values = state;
+            }
+
+            // Generate the dependencies.
+            tracing::debug_span!("generate dependencies")
+                .in_scope(|| prover.machine().generate_dependencies(&mut records, &opts));
+
+            // Defer events that are too expensive to include in every shard.
+            for record in records.iter_mut() {
+                deferred.append(&mut record.defer());
+            }
+
+            // See if any deferred shards are ready to be commited to.
+            let mut _deferred = deferred.split(false, opts.split_opts);
+
+            // Update the public values & prover state for the shards which do not contain "cpu events"
+            // before committing to them.
+            state.execution_shard += 1;
+
+            records_tx.send(records).unwrap();
+
+            processed_checkpoints += 1;
+        }
+    });
+}
+
+fn process_deferred(
+    program: &Program,
+    checkpoint: Checkpoint,
+    mut state: PublicValues<u32, u32>,
+    opts: SP1CoreOpts,
+    records_tx: SyncSender<Vec<ExecutionRecord>>,
+    deferred: &mut ExecutionRecord,
+) {
+    tracing::debug_span!("phase 1 record generator").in_scope(|| {
+        // Trace the checkpoint and reconstruct the execution records.
+        let (mut records, _) = tracing::debug_span!("trace checkpoint")
+            .in_scope(|| trace_checkpoint(program.clone(), checkpoint, opts));
+
+        // Update the public values & prover state for the shards which contain "cpu events".
+        for record in records.iter_mut() {
+            // state.shard += 1;
+            state.execution_shard = record.public_values.execution_shard;
+            state.start_pc = record.public_values.start_pc;
+            state.next_pc = record.public_values.next_pc;
+            record.public_values = state;
+        }
+
+        // See if any deferred shards are ready to be commited to.
+        let mut deferred = deferred.split(true, opts.split_opts);
+
+        // Update the public values & prover state for the shards which do not contain "cpu events"
+        // before committing to them.
+
+        for record in deferred.iter_mut() {
+            state.shard += 1;
+            state.previous_init_addr_bits = record.public_values.previous_init_addr_bits;
+            state.last_init_addr_bits = record.public_values.last_init_addr_bits;
+            state.previous_finalize_addr_bits = record.public_values.previous_finalize_addr_bits;
+            state.last_finalize_addr_bits = record.public_values.last_finalize_addr_bits;
+            state.start_pc = state.next_pc;
+            record.public_values = state;
+        }
```

Comment:
Is the state a local worker state? If yes it probably should be renamed for clarity.